source/simulator/solver_schemes.cc

```diff
@@ -0,0 +1,744 @@
+        *initial_residual = compute_initial_stokes_residual();
+      }
+
+    const double current_residual = solve_stokes();
+
+    current_linearization_point.block(introspection.block_indices.velocities)
+      = solution.block(introspection.block_indices.velocities);
+
+    if (introspection.block_indices.velocities != introspection.block_indices.pressure)
+      current_linearization_point.block(introspection.block_indices.pressure)
+        = solution.block(introspection.block_indices.pressure);
+
+    if (parameters.include_melt_transport)
+      {
+        // Note that the compaction pressure is in the fluid pressure block
+        // and will therefore be updated as well.
+        const unsigned int fluid_velocity_block = introspection.variable("fluid velocity").block_index;
+        const unsigned int fluid_pressure_block = introspection.variable("fluid pressure").block_index;
+        current_linearization_point.block(fluid_velocity_block) = solution(fluid_velocity_block);
+        current_linearization_point.block(fluid_pressure_block) = solution.block(fluid_pressure_block);
+      }
+
+    if ((initial_residual != NULL) && (*initial_residual > 0))
+      return current_residual / *initial_residual;
+
+    return 0.0;
+  }
+
+
+
+  template <int dim>
+  void Simulator<dim>::solve_IMPES ()
+  {
+    assemble_and_solve_temperature();
+    assemble_and_solve_composition();
+
+    // the Stokes matrix depends on the viscosity. if the viscosity
+    // depends on other solution variables, then after we need to
+    // update the Stokes matrix in every time step and so need to set
+    // the following flag. if we change the Stokes matrix we also
+    // need to update the Stokes preconditioner.
+    if (stokes_matrix_depends_on_solution() == true)
+      rebuild_stokes_matrix = rebuild_stokes_preconditioner = true;
+    assemble_and_solve_stokes();
+
+    if (parameters.run_postprocessors_on_nonlinear_iterations)
+      postprocess ();
+
+    return;
+  }
+
+
+
+  template <int dim>
+  void Simulator<dim>::solve_Stokes_only ()
+  {
+    double initial_stokes_residual = 0.0;
+
+    const unsigned int max_nonlinear_iterations =
+      (pre_refinement_step < parameters.initial_adaptive_refinement)
+      ?
+      std::min(parameters.max_nonlinear_iterations,
+               parameters.max_nonlinear_iterations_in_prerefinement)
+      :
+      parameters.max_nonlinear_iterations;
+    do
+      {
+        if ((stokes_matrix_depends_on_solution() == true)
+            ||
+            (parameters.prescribed_velocity_boundary_indicators.size() > 0))
+          rebuild_stokes_matrix = true;
+        if (stokes_matrix_depends_on_solution() == true)
+          rebuild_stokes_preconditioner = true;
+
+        const double relative_stokes_residual =
+          assemble_and_solve_stokes(nonlinear_iteration == 0, &initial_stokes_residual);
+
+        pcout << "      Relative Stokes residual after nonlinear iteration " << nonlinear_iteration+1
```

Comment:
Not sure if it should be changed in this pull request or in a follow up, but since it is about unification, I will bring it up here anyway.  

I think we should also unify these outputs, so that it is easier to grep them. Some for example start with 'Total relative ... 'and others with 'Relative ... '. This could then for example become something like 'Nonlinear residual (Stokes) for iteration ... ' or 'Nonlinear residual (Total) for iteration ...' etc. And then always for every timestep end with Final residual for timestep ...